### ISSUE TITLE
Trouble following packaging libraries tutorial

### ISSUE THREAD

**Issue by PaperStrike** [pypa/packaging-problems#459]
Windows 10

Python 3.9.2

21.0.1

https://packaging.python.org/tutorials/packaging-projects/#generating-distribution-archives

Honestly, I am facing with **many** problems.

The first one, is that I can't install `wheel` package.
```
for lineno, line in enumerate(fp, start=1):
UnicodeDecodeError: 'gbk' codec can't decode byte 0xba in position 122: illegal multibyte sequence
```

The second one, is that I can't run `python -m build` successfully. Every time I run it, it raise an Error:
```
raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\Users\\WangH\\AppData\\Local\\Temp\\build-env-014z0ki3\\Scripts\\python.exe', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
```

**However**, the first problem is solved by running `pip install wheel` in another path; the second is solved by removing Chinese characters in `setup.cfg`. I later found the first problem can **also** be solved by the method to the second.

But then, the build still fails,

```
raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['C:\\Users\\WangH\\AppData\\Local\\Temp\\build-env-puws9xkf\\Scripts\\python.exe', '-Im', 'pip', 'install', '--use-pep517', '--no-warn-script-location', '-r', 'C:\\Users\\WangH\\AppData\\Local\\Temp\\build-reqs-prks8sf_.txt']' returned non-zero exit status 1.
```

It seems like these packages haven't fully support UTF-8 encoding.

**Comment by PaperStrike** [comment]
The first and second problem may be located here: (Line 406 in `distutils\dist.py`)
![image](https://user-images.githubusercontent.com/22674396/109334654-f6c06500-789b-11eb-814f-5d6977499777.png)

It seems like all package packaged by `build` will run into this, and failed to decode as UTF-8 eventually, since it use `read` without `encoding`, `read` calls `open` with `encoding`, and `open` use uncertain encoding to decode when `encoding` is None. (`gbk` in my case)

![image](https://user-images.githubusercontent.com/22674396/109335107-8d8d2180-789c-11eb-873c-5cc567a314b8.png)
![image](https://user-images.githubusercontent.com/22674396/109335502-1906b280-789d-11eb-9838-3136c707804a.png)

I failed to locate other problems by myself, they seems to relate to the virtual environment. They are gone when I use `-n` in `python -m build` .

**Comment by layday** [comment]
This appears to be https://github.com/pypa/pip/issues/8931.

**Comment by layday** [comment]
Setting `PYTHONUTF8` to `1` in your environment would probably fix this in the meantime, or you could revert to using `setup.py` for package metdata.

**Comment by henryiii** [comment]
> or you could revert to using

Isn't this blocking the installs of wheel and other packages? Also, not sure why these installing from source instead of the wheel.

**Comment by layday** [comment]
I don't think they are being installed from source.  pip is attempting to read the `setup.cfg` residing in the working directory; not that of the source distributions it's installing.

**Comment by henryiii** [comment]
Ahh, that's why the mention of working in another directory.

Could you use:

```ini
long_description = file: README.rst
```

And put the special characters in that file? (name it whatever you want)

**Comment by henryiii** [comment]
Oops, nevermind, forgot that it's just unable to read "normal" characters due to the non-normal encoding. Ignore my last comment.

**Comment by PaperStrike** [comment]
Thank you all for the quick reply. And yes, setting `PYTHONUTF8` to `1` do fix the problem. 

But when I run `python -m build` without `-n`, the last problem still exists.

```
> python -m build
ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.
Traceback (most recent call last):
  File "C:\Users\WangH\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\WangH\AppData\Local\Programs\Python\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\WangH\AppData\Roaming\Python\Python39\site-packages\build\__main__.py", line 214, in <module>
    main(sys.argv[1:], 'python -m build')
  File "C:\Users\WangH\AppData\Roaming\Python\Python39\site-packages\build\__main__.py", line 206, in main
    build_package(args.srcdir, outdir, distributions, config_settings, not args.no_isolation, args.skip_dependencies)
  File "C:\Users\WangH\AppData\Roaming\Python\Python39\site-packages\build\__main__.py", line 94, in build_package
    _build_in_isolated_env(builder, outdir, distributions, config_settings)
  File "C:\Users\WangH\AppData\Roaming\Python\Python39\site-packages\build\__main__.py", line 56, in _build_in_isolated_env
    env.install(builder.build_dependencies)
  File "C:\Users\WangH\AppData\Roaming\Python\Python39\site-packages\build\env.py", line 168, in install
    subprocess.check_call(cmd)
  File "C:\Users\WangH\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['C:\\Users\\WangH\\AppData\\Local\\Temp\\build-env-yrqxraqf\\Scripts\\python.exe', '-Im', 'pip', 'install', '--use-pep517', '--no-warn-script-location', '-r', 'C:\\Users\\WangH\\AppData\\Local\\Temp\\build-reqs-yb6fz33g.txt']' returned non-zero exit status 1.
```

I copied the txt file in a special way and run the command `python -Im pip install --use-pep517 balabalabala...` in console, everything works fine. It must have something to do with the virtual environment.

I am willing to test when needed.

**Comment by layday** [comment]
Do you have `PIP_USER` in your environment or do you have a `pip.conf` - what is the output of `pip config list`?

**Comment by PaperStrike** [comment]
> the output of `python -m pip config list`
```
freeze.user='yes'
global.index-url='https://pypi.tuna.tsinghua.edu.cn/simple'
install.user='yes'
list.user='yes'
uninstall.user='yes'
```

**Comment by layday** [comment]
The user setting is inherited by pip in the build environment. Performing a user install inside a virtual environment (with default settings) is an error.  This goes for any virtual environment - not just the ones created by `build`.

**Comment by PaperStrike** [comment]
It's quite weird that I didn't configure these on purpose. I reinstalled python just before issuing to test. What I memorized is the index url I configured since PyPI is very slow in my location.

Before I reinstalled, I use the Windows-App version python. Probably these user settings are inherited from it.

Will test later, been a bit busy today.

**Comment by PaperStrike** [comment]
Setting `PIP_USER` to 0 fixed the last problem. But again, the previous problem appared 😂, same `UnicodeDecodeError` when installing `wheel` in virtual environment. It seems `PYTHONUTF8` will not affect the subprocess created in it.

**Comment by layday** [comment]
Are you sure that `PYTHONUTF8` is set? `build` does not discard the enclosing environment.

**Comment by PaperStrike** [comment]
I am sure. No matter it is set in current session or in system environment, `gbk` is still the one to decode `setup.cfg`, in may case.

**Comment by PaperStrike** [comment]
@layday I added a line 168 to ensure and noticed there is a line 160 saying `-I` or `-E` stuff.

![env.py 157-169](https://user-images.githubusercontent.com/22674396/109447634-14015900-7a7f-11eb-8c5d-f8173b6ded72.png)

According to `python -h`, `PYTHONUTF8` will be ignored in this case.

![-I and -E in python](https://user-images.githubusercontent.com/22674396/109447782-6e9ab500-7a7f-11eb-82c9-88ff02e2cf49.png)

**Comment by PaperStrike** [comment]
Use `PYTHONUTF8` environment variable or `-X utf8` option when running `python -m build`, **and** edit `env.py` like this:

![utf-8 env.py](https://user-images.githubusercontent.com/22674396/109449035-6728db00-7a82-11eb-9f04-e052b7ed5ecf.png)

will fix all the problems. (As user mode problem has started being discussed in a more proper issue.)

If only edit `env.py` in the way above, some other `UnicodeDecodeError` will appear and I haven't had time to locate.

**Comment by layday** [comment]
Ah yes, good catch. I suppose your only remaining option is to use `setup.py` until this particular distutils holdover is removed from pip or something about it is done in build.  I don't think we should override the system encoding in build but you might want to open an issue on the build tracker so that other developers are able to chime in.

**Comment by henryiii** [comment]
The env variable and flag was added in Python 3.7. I think there’s likely to be a fix from Pip sort of soonish.

**Comment by PaperStrike** [comment]
Agreed.

Problems here have found their homes and are likely to be fixed soon.
Thanks for all your support!

Will close this after the fixes.

**Comment by layday** [comment]
The Unicode issues should have been fixed in pip 21.2, so I'll go ahead and close this now.  Thanks for the detailed report and investigation.
